lib/spack/spack/package_prefs.py

```diff
@@ -252,5 +253,76 @@ def is_spec_buildable(spec):
     return allpkgs[spec.name]['buildable']
 
 
+def get_package_dir_permissions(spec):
+    """Return the permissions configured for the spec.
+
+    Include the GID bit if group permissions are on. This makes the group
+    attribute sticky for the directory."""
+    perms = get_package_permissions(spec)
+    if perms & stat.S_IRWXG:
+        perms |= stat.S_ISGID
+    return perms
+
+
+def get_package_permissions(spec):
+    """Return the permissions configured for the spec"""
+    allpkgs = get_packages_config()
+
+    # Get readable permissions level
+    if spec.name in allpkgs and 'readable' in allpkgs[spec.name]:
+        readable = allpkgs[spec.name]['readable']
+    elif 'all' in allpkgs and 'readable' in allpkgs['all']:
+        readable = allpkgs['all']['readable']
+    else:
+        readable = 'world'
+    if readable not in ('user', 'group', 'world'):
+        raise ConfigError("Invalid 'readable' permissions %s.\n" % spec.name +
+                          "      Valid values are 'user', 'group', 'world'")
+
+    # Get writable permissions level
+    if spec.name in allpkgs and 'writable' in allpkgs[spec.name]:
+        writable = allpkgs[spec.name]['writable']
+    elif 'all' in allpkgs and 'writable' in allpkgs['all']:
+        writable = allpkgs['all']['writable']
+    else:
+        writable = 'user'
+    if readable not in ('user', 'group', 'world'):
+        raise ConfigError("Invalid 'readable' permissions %s.\n" % spec.name +
+                          "      Valid values are 'user', 'group', 'world'")
+
+    # rwx permissions set by build_system
+    # Only ugo granularity configurable
+    perms = stat.S_IRWXU
+    if readable in ('world', 'group'):  # world includes group
+        perms |= stat.S_IRGRP | stat.S_IXGRP
+    if readable == 'world':
+        perms |= stat.S_IROTH | stat.S_IXOTH
+    if writable in ('world', 'group'):
+        if readable == 'user':
+            raise ConfigError('Writable permissions may not be more' +
+                              ' permissive than readable permissions.\n' +
+                              '      Violating package is %s' % spec.name)
```

Comment:
I'd just say `In package '%s': ... ` instead of "violating package" 👮 🚔 .

Also more importantly, these should be `ConfigFileError`s, because that will give the user file/line context so they know what to fix.

Reply:
@tgamblin It doesn't look like `ConfigFileError`s give file/line context, are you sure you're thinking of the right error?

Reply:
Sorry, I meant ConfigFormatError, but it’s not quite right either as it’s meant for schema issues.

Look at ConfigFormatError._get_mark — The dicts, lists, strings, ints, etc that come out of the YAML parser have _start_mark and _end_mark attributes with file/line info. This should grab them and produce a useful error message. Probably there should be another error type alongside ConfigFormatError, and the error finding logic should be generalized so it’s useful for both and not written twice.